### ISSUE TITLE
--skip-simulation does not skip simulation

### ISSUE THREAD

**Issue by Philogy** [foundry-rs/foundry#6825]
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (24abca6 2024-01-16T00:26:29.551032000Z)

### What command(s) is the bug in?

forge script --skip-simulation

### Operating System

macOS (Apple Silicon)

### Describe the bug

Often there are things that my script depends on that foundry is unable to simulate correctly (self-destruct reinit, hot/cold storage gas cost).

Whether `--skip-simulation` is provided or not `forge script` will simulate the script first before attempting to broadcast it. Meaning that if it's incorrectly failing in the simulation I'm unable to broadcast transactions using foundry.

The `--skip-simulation` flag does however get the script runner to exclaim "SKIPPING ON CHAIN SIMULATION." after it has successfully _ran the simulation_ 😅
![image](https://github.com/foundry-rs/foundry/assets/21957732/a6c0dea2-53ed-4e70-9ebb-6a6083e18e16)

**Comment by mattsse** [comment]
thanks for flagging, possible regression, fixing asap

**Comment by mattsse** [comment]
do you have a repro for this?

**Comment by Sabnock01** [comment]
Had some previous discussion in #5776

**Comment by Philogy** [comment]
Yeah based on @Evalir's comment here: https://github.com/foundry-rs/foundry/issues/5776#issuecomment-1867287499 it does seem like this _intentional_ and kind of unavoidable based on how scripts currently work.

Unfortunately, this does mean that `forge script` is unusable for me in many use-cases. Specifically because of the simulation being in 1 tx (meaning gas accounting for warm/cold storage slots will be off and self-destruct / reinit doesn't work).

Not sure how you'd fix this exactly, maybe have a separate environment in which the script contract is run and the actual simulation itself, any external CALLs made in one are captured and emitted as transactions in the actual environment?

**Comment by Philogy** [comment]
Here's a minimal [repro](https://gist.github.com/Philogy/90e0a294b2587162ab14027c8f2fe14a)

This seems like a niche case but there are others like self-destruct, reinit or other weird dependencies like gas-price that would cause the same issue

**Comment by plotchy** [comment]
copying good background provided by @evalir from #5776 :
> --skip-simulation avoids using the provided URL to do gas estimation. It does not prevent foundry from simulating the script locally. The latter is not possible to disable, as we must simulate the script to gather all the transactions to broadcast. If the local simulation fails, the whole script fails.

much of my experience with wanting this feature to avoid local sim was around similar edge cases. On-chain randomness where the randomness is seeded by things like gas, gas-limit, coinbase, etc that is difficult to match exactly in a forge script.

**Comment by Xenoset77** [comment]
Can we get an ETA on this?

**Comment by jenpaff** [comment]
this is a thorough re-design of forge script, putting into backlog

**Comment by mshakeg** [comment]
this seems to still be an issue

**Comment by 0xMeir** [comment]
Still an issue

**Comment by maxim-inj** [comment]
wtf................

**Comment by grasphoper** [comment]
Still an issue. @mattsse could you help?

**Comment by nicholaspai** [comment]
This would be a very helpful workaround when developing on EVM's that have precompile contracts, which currently `forge` is unable to call/staticcall (I'm not sure why) and complains that the simulation would revert `call to non-contract address...`

**Comment by mablr** [comment]
Closing because I'm no longer able to reproduce this as reported on current master (https://github.com/foundry-rs/foundry/commit/bbe48859195f6adad8dbf68162ac1f06ad9d58d9). The exact repro broadcasts successfully and skips the on-chain simulation. Forge must still execute `run()` locally to discover transactions; that is expected imo.
